lib/full-site-editing/block-templates.php

```diff
@@ -93,18 +113,62 @@ function _gutenberg_get_template_files( $template_type ) {
 				// Subtract ending '.html'.
 				-5
 			);
-			$template_files[] = array(
+
+			$new_template_item = array(
 				'slug'  => $template_slug,
 				'path'  => $template_file,
 				'theme' => $theme_slug,
 				'type'  => $template_type,
 			);
+
+			if ( 'wp_template_part' === $template_type ) {
+				$template_files[] = _gutenberg_conditionally_add_template_part_section_info(
+					$new_template_item,
+					$template_part_data
+				);
+			} else {
+				$template_files[] = $new_template_item;
+			}
 		}
 	}
 
 	return $template_files;
 }
 
+/**
+ * Attempts to read the theme.json and return the 'template-parts' field.
+ *
+ * @return array Template part data from theme.json or empty array if not found.
+ */
+function _gutenberg_get_template_part_info_from_theme_json() {
+	if ( is_readable( locate_template( 'experimental-theme.json' ) ) ) {
+		$theme_json = file_get_contents( locate_template( 'experimental-theme.json' ) );
+		$data       = json_decode(
+			$theme_json,
+			true
+		);
+	}
+	if ( isset( $data['template-parts'] ) ) {
+		return $data['template-parts'];
+	}
+	return array();
+}
+
+/**
+ * Attempts to add the template part's section information to the input template.
+ *
+ * @param array $template_info Template to add information to (requires 'type' and 'slug' fields).
+ * @param array $theme_data Template part information from theme.json.
+ *
+ * @return array Template.
+ */
+function _gutenberg_conditionally_add_template_part_section_info( $template_info, $theme_data ) {
```

Comment:
>I've seen a couple of usages of leading underscore through Gutenberg, but I don't think it's documented in WP coding conventions. Do we really need it here?

I noticed some of the other functions in this file use the leading underscore as well.  From inspection I am assuming helper functions only intended to be called in this file have the leading underscore, while more robust functions also used elsewhere do not.

>Also I think we could drop conditionally from the name of the function since I don't think it helps much with readability.

👍  Yeah, after refactoring that no longer seems necessary.